apps/indexer/src/api/services/treasury/treasury-provider-factory.ts

```diff
@@ -0,0 +1,37 @@
+import { env } from "@/env";
+import axios from "axios";
+import { DefiLlamaProvider } from "./providers/defillama–provider";
+import { DuneProvider } from "./providers/dune-provider";
+import { TreasuryService } from "./treasury.service";
+import { assets } from "@/api/controllers";
+import { OpenAPIHono as Hono } from "@hono/zod-openapi";
+
+/**
+ * Creates a treasury provider and registers API routes
+ * Providers fetch data on-demand
+ * @param app - The Hono app instance
+ * @returns void
+ * @throws Error if no treasury provider is configured
+ */
+export function createTreasuryProvider(app: Hono) {
+  if (env.TREASURY_PROVIDER_PROTOCOL_ID && env.DEFILLAMA_API_URL) {
+    const axiosClient = axios.create({
+      baseURL: env.DEFILLAMA_API_URL,
+    });
+    const defiLlamaProvider = new DefiLlamaProvider(
+      axiosClient,
+      env.TREASURY_PROVIDER_PROTOCOL_ID,
+    );
+    const treasuryService = new TreasuryService(defiLlamaProvider);
+    assets(app, treasuryService);
+  } else if (env.DUNE_API_URL && env.DUNE_API_KEY) {
+    const axiosClient = axios.create({
+      baseURL: env.DUNE_API_URL,
+    });
+    const duneProvider = new DuneProvider(axiosClient, env.DUNE_API_KEY);
+    const treasuryService = new TreasuryService(duneProvider);
+    assets(app, treasuryService);
+  } else {
+    throw new Error("Not enough variables to create a treasury provider");
```

Comment:
what if we don't want to support this endpoint for some DAOs? (e.g, how would Obol work if its not supported by any of those providers?)